src/VisualStudio/Core/Def/Utilities/ClipboardHelpers.cs

```diff
@@ -0,0 +1,152 @@
+        /// For cases where the clipboard data is retrieved in a performance critical path this 
+        /// should be used to avoid the overhead that WPF/WinForms adds for the clipboard APIs.
+        /// If performance is less of a concern (for example responding directly to a user paste
+        /// command instead of VS activation) then it is recommended to use paste command handlers
+        /// or the clipboard API directly. 
+        /// </summary>
+        public static string? GetTextNoRetry()
+        {
+            if (OleGetClipboard(out var dataObject) != VSConstants.S_OK || dataObject is null)
+            {
+                return null;
+            }
+
+            foreach (var format in TextFormats)
+            {
+                if (dataObject.QueryGetData(new[] { format }) == VSConstants.S_OK)
+                {
+                    return GetData(dataObject, format);
+                }
+            }
+
+            return null;
+        }
+
+        // 
+        // The rest of the methods are derived from the WPF implementation of getting
+        // data from an OLE IDataObject. We use our own logic because there are built in 
+        // mechanisms within WPF that result in retries on getting clipboard data which
+        // is undesirable for paths that check the clipboard data in a hot path. 
+        // See https://github.com/dotnet/wpf/blob/212f376fbca58bf2970964610c426ee05e633872/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/dataobject.cs
+        // for information on how dataobject is used in WPF. This code only supports
+        // a subset of the functionality for now because it is limited to text.
+        //
+        private static string? GetData(IDataObject dataObject, FORMATETC format)
+        {
+            var af = new FORMATETC[1];
+            af[0] = format;
+            var sm = new STGMEDIUM[1];
+            dataObject.GetData(af, sm);
+            var medium = sm[0];
+
+            try
+            {
+                if (medium.tymed != 1) // TYMED_HGLOBAL
+                {
+                    return null;
+                }
+
+                return ReadStringFromHandle(medium.unionmember, format.cfFormat == CF_UNICODETEXT);
+            }
+            finally
+            {
+                ReleaseStgMedium(ref medium);
+            }
+        }
+
+        private static FORMATETC CreateFormatEtc(ushort format)
+            => new FORMATETC
+            {
+                cfFormat = format,
+                ptd = IntPtr.Zero,
+                dwAspect = (uint)DVASPECT.DVASPECT_CONTENT,
+                lindex = -1,
+                tymed = (uint)TYMED.TYMED_HGLOBAL
+            };
+
+        private static unsafe string? ReadStringFromHandle(IntPtr handle, bool unicode)
+        {
+            string? stringData = null;
+            IntPtr ptr;
+            object handleRefObj = new();
+
+            ptr = Win32GlobalLock(new HandleRef(handleRefObj, handle));
```

Comment:
Is there a reason to be using HandleRef here versus just GlobalLock directly being an IntPtr? It looks like the clipboard code you're copying from was trying to use that to keep some other object alive, but it's not doing anything here?

Reply:
Would that just be putting a global lock on the handle at that point instead of associating a ref object with it?